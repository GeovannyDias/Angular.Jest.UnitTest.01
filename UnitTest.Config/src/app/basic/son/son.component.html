<h5>Son Component</h5>
<code>{{ client | json }}</code>
<br><br>

<button *ngIf="client" class="btn btn-outline-primary m-2" (click)="onChangeId(5)" data-test="btn-change-id">Change ID</button>
<button *ngIf="client" class="btn btn-outline-danger m-2" (click)="onDelete()" data-test="btn-delete">Delete Client</button>