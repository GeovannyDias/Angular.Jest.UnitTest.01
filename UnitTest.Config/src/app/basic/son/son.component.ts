import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { ClientI } from '../interfaces/client.interface';

@Component({
  selector: 'app-son',
  templateUrl: './son.component.html',
  styleUrls: ['./son.component.scss']
})
export class SonComponent implements OnInit {
  @Input() client?: ClientI;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdate = new EventEmitter<ClientI>();

  constructor() { }

  ngOnInit(): void {
  }

  onChangeId(id: number) {
    // if (this.client) this.client.id = id; // No usar pasa por referencia y se actualiza en el fatherComponent
    if (this.client) {
      this.client = { ...this.client, id };
      this.onClientUpdate.emit({ ...this.client });
    }
  }

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

}
