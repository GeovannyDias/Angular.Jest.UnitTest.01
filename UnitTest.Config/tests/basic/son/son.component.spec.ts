import { ComponentFixture, TestBed } from '@angular/core/testing';

import { SonComponent } from '../../../src/app/basic/son/son.component';

describe('SonComponent', () => {
  let component: SonComponent;
  let fixture: ComponentFixture<SonComponent>;
  let compiled: HTMLElement; //

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [SonComponent]
    })
      .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(SonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement; //
    jest.clearAllMocks(); // Limpia todas las funciones de jest que han sido llamadas (reiniciar)
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  // ---------------------------------------------------------------------------

  // Optional

  test('should do match to snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should not show buttos, if dont exist client', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0);
  });

  test('should show buttos, if exist client', () => {
    component.client = { id: 1, name: 'Geovanny' };
    fixture.detectChanges();
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  });

  // Optional

  test('should do match to snapshot, if exist client', () => {
    component.client = { id: 1, name: 'Geovanny' };
    fixture.detectChanges();
    expect(compiled).toMatchSnapshot();
  });

  test('should emit onDeleteClient event with button of Delete Client', () => {
    component.client = { id: 1, name: 'Geovanny' };
    fixture.detectChanges();
    jest.spyOn(component.onDeleteClient, 'emit'); // Esta espiando en el componente u objeto (emit)
    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    // console.log('btnDelete1:', btnDelete?.innerHTML);
    // console.log('btnDelete2:', btnDelete?.textContent);
    btnDelete?.dispatchEvent(new Event('click'));
    // expect(component.onDeleteClient.emit).toHaveBeenCalledWith('Hello World');
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  test('should emit onClientUpdate event with button of "Change ID"', () => {
    component.client = { id: 1, name: 'Geovanny' };
    fixture.detectChanges();
    jest.spyOn(component.onClientUpdate, 'emit'); // Esta espiando en el componente u objeto (emit)
    const btnChangeId = compiled.querySelector('[data-test=btn-change-id]');
    // btnChangeId?.dispatchEvent(new Event('click')(5));
    btnChangeId?.dispatchEvent(new Event('click'));

    // expect(component.onClientUpdate.emit).toHaveBeenCalled();
    expect(component.onClientUpdate.emit).toHaveBeenCalledWith({ id: 5, name: 'Geovanny' });
  });

  test('should emit onChangeClient with specific ID, if there is a client', () => {  
    jest.spyOn(component.onClientUpdate, 'emit'); // Se convierte en una funcion que es llamada, se puede verificar las veces que se llam√≥, etc.
    component.onChangeId(10);
    expect(component.onClientUpdate.emit).not.toHaveBeenCalled(); // SI NO hay un cliente no se ejecuta la condicion de onChangeId(10)
    
    component.client = { id: 1, name: 'Geovanny' };
    fixture.detectChanges();
    component.onChangeId(10);

    expect(component.onClientUpdate.emit).toHaveBeenCalledWith({ id: 10, name: 'Geovanny' });
  });



});
