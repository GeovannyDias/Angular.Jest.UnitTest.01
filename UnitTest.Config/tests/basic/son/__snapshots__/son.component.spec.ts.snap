// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SonComponent should do match to snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Son Component
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;

exports[`SonComponent should do match to snapshot, if exist client 1`] = `
<div
  id="root4"
>
  <h5>
    Son Component
  </h5>
  <code>
    {
  "id": 1,
  "name": "Geovanny"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary m-2"
    data-test="btn-change-id"
  >
    Change ID
  </button>
  
  <button
    class="btn btn-outline-danger m-2"
    data-test="btn-delete"
  >
    Delete Client
  </button>
  
</div>
`;
